tests/unit/azure/test_credentials.py

```diff
@@ -0,0 +1,337 @@
+import io
+import logging
+import re
+from unittest.mock import MagicMock, create_autospec
+
+import pytest
+import yaml
+from databricks.labs.blueprint.installation import MockInstallation
+from databricks.labs.blueprint.tui import MockPrompts
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import NotFound, ResourceDoesNotExist
+from databricks.sdk.errors.platform import InvalidParameterValue
+from databricks.sdk.service.catalog import (
+    AwsIamRole,
+    AzureManagedIdentity,
+    AzureServicePrincipal,
+    StorageCredentialInfo,
+    ValidateStorageCredentialResponse,
+)
+from databricks.sdk.service.workspace import GetSecretResponse
+
+from databricks.labs.ucx.assessment.azure import (
+    AzureServicePrincipalCrawler,
+    AzureServicePrincipalInfo,
+)
+from databricks.labs.ucx.azure.access import (
+    AzureResourcePermissions,
+    StoragePermissionMapping,
+)
+from databricks.labs.ucx.azure.credentials import (
+    ServicePrincipalMigration,
+    ServicePrincipalMigrationInfo,
+    StorageCredentialManager,
+)
+
+
+@pytest.fixture
+def ws():
+    state = {"/Users/foo/.ucx/config.yml": yaml.dump({'version': 2, 'inventory_database': 'ucx'})}
+
+    def download(path: str) -> io.StringIO:
+        if path not in state:
+            raise NotFound(path)
+        return io.StringIO(state[path])
+
+    ws_mock = create_autospec(WorkspaceClient)
+    ws_mock.config.host = 'https://localhost'
+    ws_mock.current_user.me().user_name = "foo"
+    ws_mock.workspace.download = download
+    return ws_mock
+
+
+def side_effect_create_storage_credential(name, azure_service_principal, comment, read_only):
+    return StorageCredentialInfo(
+        name=name, azure_service_principal=azure_service_principal, comment=comment, read_only=read_only
+    )
+
+
+def side_effect_validate_storage_credential(storage_credential_name, url, read_only):  # pylint: disable=unused-argument
+    if "overlap" in storage_credential_name:
+        raise InvalidParameterValue
+    if read_only:
+        response = {"isDir": True, "results": [{"message": "", "operation": "READ", "result": "PASS"}]}
+        return ValidateStorageCredentialResponse.from_dict(response)
```

Comment:
```suggestion
        return ValidateStorageCredentialResponse(is_dir=True, results=[...])
```
don't use dictionaries, where you can use typed dataclasses

Reply:
fixed